### PR TITLE
[maskedtensor] Advanced semantics [4/4]

### PR DESCRIPTION
**Pull request by george-qi**
Continuation of https://github.com/pytorch/tutorials/pull/2052. 

Restructured the PR stack from https://github.com/pytorch/tutorials/pull/2042 

The new format will be:

PR #1: Overview Tutorial
- [Old overview](https://github.com/pytorch/tutorials/pull/2042)
- Why is MaskedTensor useful
    - [Distinguish 0 and NaN gradient](https://github.com/pytorch/tutorials/pull/2044)
    - [Safe softmax tutorial](https://github.com/pytorch/tutorials/pull/2045)
    - [Missing nan ops](https://github.com/pytorch/tutorials/pull/2046)

PR #2: [Sparsity](https://github.com/pytorch/tutorials/pull/2043)

PR #3: [Adagrad sparse semantics](https://github.com/pytorch/tutorials/pull/2047), i.e. what does MaskedTensor make easier -- this one

PR #4: Advanced semantics, e.g. NumPy differences and reduction semantics from [Old overview](https://github.com/pytorch/tutorials/pull/2042)

### CODE REVIEW

File: prototype_source/maskedtensor_advanced_semantics.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+Note that the mask is `mt0.get_mask() & mt1.get_mask()` since :class:`MaskedTensor`'s mask is the inverse of NumPy's.
+
+.. _reduction-semantics:
+
+Reduction semantics
+-------------------
+
+Recall in `MaskedTensor's Overview tutorial <https://pytorch.org/tutorials/prototype/maskedtensor_overview.html>`__
+we discussed "Implementing missing torch.nan* ops". Those are examples of reductions -- operators that remove one
+(or more) dimensions from a Tensor and then aggregate the result. In this section, we'll use reduction semantics
```
Comment (svekars) [inline prototype_source/maskedtensor_advanced_semantics.rst:88]:
```suggestion
(or more) dimensions from a Tensor and then aggregate the result. In this section, we will use reduction semantics
```

File: prototype_source/maskedtensor_advanced_semantics.rst [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    >>> npm1
+    masked_array(
+      data=[[10.0, 11.0, 12.0, --, --],
+            [--, --, 17.0, 18.0, 19.0]],
+      mask=[[False, False, False,  True,  True],
+            [ True,  True, False, False, False]],
+      fill_value=1e+20,
+      dtype=float32)
+
+Now let's try addition:
```
Comment (svekars) [inline prototype_source/maskedtensor_advanced_semantics.rst:165]:
```suggestion
Now, let's try addition:
```

File: prototype_source/maskedtensor_advanced_semantics.rst [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+                mask=[False, False, False, False, False],
+          fill_value=1e+20,
+                dtype=float32)
+
+Sum and addition should clearly be associative, but with NumPy's semantics, they are allowed to not be,
+which can certainly be confusing for the user.
+
+:class:`MaskedTensor`, on the other hand, will simply not allow this operation since `mask0 != mask1`.
+That being said, if the user wishes, there are ways around this
+(e.g. filling in the MaskedTensor's undefined elements with 0 values using :func:`to_tensor` like shown below),
```
Comment (svekars) [inline prototype_source/maskedtensor_advanced_semantics.rst:183]:
```suggestion
(for example, filling in the MaskedTensor's undefined elements with 0 values using :func:`to_tensor` like shown below),
```

File: prototype_source/maskedtensor_advanced_semantics.rst [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    >>> torch.amax(mt, 1)
+    MaskedTensor(
+      [  1.0000,   7.0000,  11.0000]
+    )
+
+Of note, the value under a masked out element is not guaranteed to have any specific value, especially if the
+row or column is entirely masked out (the same is true for normalizations).
+For more details on masked semantics, you can find this `RFC <https://github.com/pytorch/rfcs/pull/27>`__.
+
+Now we can revisit the question: why do we enforce the invariant that masks must match for binary operators?
```
Comment (svekars) [inline prototype_source/maskedtensor_advanced_semantics.rst:139]:
```suggestion
Now, we can revisit the question: why do we enforce the invariant that masks must match for binary operators?
```

File: prototype_source/maskedtensor_advanced_semantics.rst (hunk 3 above)
Comment (svekars) [inline prototype_source/maskedtensor_advanced_semantics.rst:183]: Please add a Conclusion section and if applicable Further Reading

**Review by jisaacso — APPROVED** [review]
LG, happy to rereview after rendering of code blocks
